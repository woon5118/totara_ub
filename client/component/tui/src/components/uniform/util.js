/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import { pick } from 'tui/util';
import { getPropDefs, getModelDef } from 'tui/vue_util';
import FormField from 'tui/components/uniform/FormField';
import { mergeListeners } from '../../js/internal/vnode';
import { charLengthProp } from '../form/form_common';

/**
 * Create a wrapper component for an input.
 *
 * The only hard requirement is that the input component works with v-model.
 *
 * @param {(object|Vue)} input Input component to wrap.
 * @param {object} [options]
 * @param {boolean} [options.passAriaLabelledby] Pass the label ID as aria-labelledby.
 * @param {boolean} [options.functional] Generate a functional component.
 * @param {boolean} [options.touchOnChange] Mark input as touched whenever it changes.
 * @returns {(object|Vue)} Vue component.
 */
export function createUniformInputWrapper(input, options = {}) {
  const functional =
    options.functional !== undefined ? options.functional : true;
  const inputProps = Object.assign({}, getPropDefs(input));
  delete inputProps.name;
  delete inputProps.validate;
  delete inputProps.validations;
  delete inputProps.id;
  delete inputProps.ariaDescribedby;
  delete inputProps.ariaInvalid;
  delete inputProps.ariaLabelledby;

  const props = Object.assign({}, inputProps, {
    id: {},

    name: {
      type: [String, Number, Array],
      required: true,
    },

    validate: Function,
    validations: [Function, Array],
    charLength: charLengthProp,
  });

  const model = getModelDef(input);
  const modelProp = (model && model.prop) || 'value';
  const modelEvent = (model && model.event) || 'input';

  return {
    functional,

    // needed for lang string loader support
    components: {
      FormField,
      Input: input,
    },

    inheritAttrs: false,

    props: props,

    render(h, context) {
      const props = functional ? context.props : this;
      const listeners = functional ? context.listeners : this.$listeners;
      const propsForInput = pick(props, Object.keys(inputProps));
      propsForInput.charLength = 'full';
      return h(FormField, {
        props: {
          name: props.name,
          validate: props.validate,
          validations: props.validations,
          charLength: props.charLength,
        },
        scopedSlots: {
          default: ({ value, update, blur, touch, labelId, attrs }) => {
            const finalProps = Object.assign({}, propsForInput);
            finalProps[modelProp] = value;
            // attrs that are actually props will automatically be moved to
            // props by Vue
            const inputAttrs = Object.assign({}, attrs);
            if (options.passAriaLabelledby) {
              inputAttrs['aria-labelledby'] = labelId;
            }
            if (props.id) {
              inputAttrs.id = props.id;
            }
            return h(
              input,
              {
                props: finalProps,
                attrs: inputAttrs,
                on: mergeListeners(listeners, {
                  [modelEvent]: value => {
                    update(value);
                    if (options.touchOnChange) touch();
                  },
                  blur,
                }),
                scopedSlots: functional ? undefined : this.$scopedSlots,
              },
              functional ? context.children : undefined
            );
          },
        },
      });
    },
  };
}
