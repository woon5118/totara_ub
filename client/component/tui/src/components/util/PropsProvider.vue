<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Simon Chester <simon.chester@totaralearning.com>
  @module tui
-->

<script>
import { cloneVNode, mergeListeners } from '../../js/internal/vnode';

/**
 * Functional component that modifies vnodes of its children, adding props and
 * event listeners.
 *
 * Note: native elements are ignored (not modified).
 *
 * Use cases:
 *   * Tightly coupled parent->child component relationships, for example
 *     <Radio>s in a <RadioGroup>.
 *   * Applying native listeners to components from renderless components
 *     (e.g. Draggable), until Vue 3 arrives and solves this.
 */
export default {
  functional: true,

  props: {
    provide: {
      type: [Function, Object],
      required: true,
    },
  },

  render(h, { props, children }) {
    return children.map(vnode => {
      vnode = cloneVNode(vnode);
      const opts = vnode.componentOptions;
      if (!opts) return vnode;
      const info = { props: opts.propsData || {} };
      const provided =
        typeof props.provide === 'function'
          ? props.provide(info)
          : props.provide;

      // set provided props...
      if (provided.props) {
        opts.propsData = Object.assign({}, provided.props, opts.propsData);
      }

      if (provided.listeners) {
        opts.listeners = mergeListeners(opts.listeners, provided.listeners);
      }

      if (provided.nativeListeners) {
        // vue event listeners go on `componentOptions.listeners`
        // native listeners go on `data.on`
        // https://github.com/vuejs/vue/blob/b9de23b1008b52deca7e7df40843e318a42f3f53/src/core/vdom/create-component.js#L166
        vnode.data.on = mergeListeners(vnode.data.on, provided.nativeListeners);
      }

      return vnode;
    });
  },
};
</script>
