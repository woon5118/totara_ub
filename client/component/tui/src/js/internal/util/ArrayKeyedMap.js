/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

/**
 * Implements has, get, set, and delete methods from Map.
 *
 * Keys must be arrays.
 */
export default class ArrayKeyedMap {
  constructor() {
    this._keys = [];
    this._values = [];
  }

  /**
   * Check if the key has a value in the map.
   *
   * @param {Array} key
   * @returns {boolean}
   */
  has(key) {
    return this._findIndex(key) !== -1;
  }

  /**
   * Get the value associated with the key, or undefined if none.
   *
   * @param {Array} key
   * @returns {*}
   */
  get(key) {
    const index = this._findIndex(key);
    return index !== -1 ? this._values[index] : undefined;
  }

  /**
   * Associate a value with a key.
   *
   * @param {Array} key
   * @param {*} value
   * @returns {ArrayKeyedMap} The current map.
   */
  set(key, value) {
    const index = this._findIndex(key);
    if (index === -1) {
      this._keys.push(Array.prototype.slice.apply(key));
      this._values.push(value);
    } else {
      this._values[index] = value;
    }
    return this;
  }

  /**
   * Remove the value associated with the key from the map.
   *
   * @param {Array} key
   * @returns {boolean} True if the key had a value, false if it did not.
   */
  delete(key) {
    const index = this._findIndex(key);
    if (index !== -1) {
      this._keys.splice(index, 1);
      this._values.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * Find the index of the key in the data array, or -1 if not found.
   *
   * @private
   * @param {Array} key
   * @returns {number} Index, or -1 if not found.
   */
  _findIndex(key) {
    if (!Array.isArray(key)) {
      throw new Error('key must be an array');
    }
    return this._keys.findIndex(x => this._keyEqual(x, key));
  }

  /**
   * Check if two keys are equal.
   *
   * Compares key array elements using SameValueZero.
   *
   * @private
   * @param {Array} a
   * @param {Array} b
   * @returns {boolean}
   */
  _keyEqual(a, b) {
    return a.length === b.length && a.every((x, i) => sameValueZero(x, b[i]));
  }
}

/**
 * Compare two values using the SameValueZero algorithm.
 *
 * {@see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness}
 *
 * @param {*} x
 * @param {*} y
 * @returns {boolean}
 */
function sameValueZero(x, y) {
  return (
    x === y ||
    (typeof x === 'number' && typeof y === 'number' && isNaN(x) && isNaN(y))
  );
}
