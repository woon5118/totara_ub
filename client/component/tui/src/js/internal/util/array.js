/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

/**
 * Remove the first instance of the specified item from the array.
 *
 * @param {array} array
 * @param {*} item
 */
export function pull(array, item) {
  const index = array.indexOf(item);
  if (index != -1) {
    array.splice(index, 1);
  }
}

/**
 * Return a copy of array filtered to only contain unique values.
 *
 * @param {array} arr
 * @returns {array}
 */
export function unique(arr) {
  return arr.filter((item, pos) => arr.indexOf(item) === pos);
}

/**
 * Create a new object composed of the selected keys of the provided object.
 *
 * @param {object} object
 * @param {array} keys
 * @return {object}
 */
export function pick(object, keys) {
  if (!Array.isArray(keys)) {
    throw new Error('keys must be an array');
  }
  const newObj = {};
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    if (key in object) {
      newObj[key] = object[key];
    }
  }
  return newObj;
}

/**
 * Splits a collection into sets, group by the result of running each value
 * through `fn`.
 *
 * @param {*} array
 * @param {*} fn
 */
export function groupBy(array, fn) {
  const result = {};
  array.forEach(x => {
    const key = fn(x);
    if (!result[key]) {
      result[key] = [];
    }
    result[key].push(x);
  });
  return result;
}

/**
 * Sort the supplied array using the supplied function to generate a sort key.
 *
 * @param {array} array
 * @param {function} fn
 * @returns {array}
 */
export function orderBy(array, fn) {
  array = array.slice();
  array.sort((a, b) => {
    const aScore = fn(a);
    const bScore = fn(b);
    if (aScore < bScore) {
      return -1;
    } else if (aScore > bScore) {
      return 1;
    } else {
      return 0;
    }
  });
  return array;
}
