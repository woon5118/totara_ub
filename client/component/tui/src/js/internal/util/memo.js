/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

/**
 * Create a memoized version of the provided function.
 *
 * By default, the first argument to the function is used as the cache key.
 * If you pass a function as `keyFn`, it will be called with the invocation's
 * arguments to generate a cache key.
 *
 * The cache key can be anything, it is not restricted to being a string.
 *
 * @param {function} fn
 * @param {function=} keyFn
 * @returns {function}
 */
export function memoize(fn, keyFn) {
  return memoizeWithMap(fn, keyFn, new Map());
}

/**
 * Create a memoized version of the provided function.
 *
 * Like `memoize()` but using a WeakMap.
 *
 * @param {function} fn
 * @param {function=} keyFn
 * @returns {function}
 */
export function memoizeWeak(fn, keyFn) {
  return memoizeWithMap(fn, keyFn, new WeakMap());
}

/**
 * Create a memoized version of the provided function.
 *
 * Like `memoize()` but using a the map you pass for key storage.
 *
 * @private
 * @param {function} fn
 * @param {function=} keyFn
 * @param {object} map Object implementing Map API (has/get/set)
 * @returns {function}
 */
function memoizeWithMap(fn, keyFn, map) {
  return function(arg) {
    const key = keyFn ? keyFn.apply(this, arguments) : arg;
    if (map.has(key)) {
      return map.get(key);
    }
    const result = fn.apply(this, arguments);
    map.set(key, result);
    return result;
  };
}

/**
 * Wrap the provided async or promise-returning function so that subsequent
 * calls get the same result, and calls while loading get the same promise
 * to avoid making multiple requests.
 *
 * @param {function} fn
 * @return {Promise}
 */
export function memoizeLoad(fn) {
  let promise;
  const memoized = () => {
    if (promise) {
      return promise;
    }

    promise = fn();
    promise.catch(() => (promise = null));
    return promise;
  };

  /* istanbul ignore next */
  if (process.env.NODE_ENV == 'test') {
    memoized._fn = fn;
    memoized._change = newFn => {
      memoize._fn = fn = newFn;
      promise = null;
    };
  }
  return memoized;
}
