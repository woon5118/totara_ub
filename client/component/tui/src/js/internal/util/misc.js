/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

let counter = 1;

/**
 * Generates an incrementing unique ID.
 *
 * This number is only unique within a page, and is not unique across page
 * loads.
 *
 * @return {number}
 */
export function uniqueId() {
  return counter++;
}

/**
 * Get a result from a value.
 *
 * If value is a function it will be called to obtain the result, otherwise
 * value will be used as-is.
 *
 * @param {*} value
 * @return {*}
 */
export function result(value) {
  if (value instanceof Function) {
    return value();
  }
  return value;
}

export function createNewEvent(eventName) {
  let event;
  if (typeof Event === 'function') {
    event = new Event(eventName);
  } else {
    event = document.createEvent('Event');
    event.initEvent(eventName, true, true);
  }
  return event;
}
