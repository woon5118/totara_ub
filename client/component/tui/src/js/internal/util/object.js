/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

/**
 * Object.prototype.hasOwnProperty wrapper.
 *
 * @param {object} obj
 * @param {string} key
 * @returns {boolean}
 */
export function hasOwnProperty(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}

const objectConstructorString = Function.prototype.toString.call(Object);

/**
 * Check if an object is a plain object (i.e. {}), not a class instance etc.
 *
 * @param {*} value
 * @returns {boolean}
 */
export function isPlainObject(value) {
  if (!value || typeof value != 'object') {
    return false;
  }
  const proto = Object.getPrototypeOf(value);
  if (proto == null) {
    return true;
  }
  const ctor = hasOwnProperty(proto, 'constructor') && proto.constructor;
  return (
    typeof ctor == 'function' &&
    ctor instanceof ctor &&
    Function.prototype.toString.call(ctor) === objectConstructorString
  );
}

/**
 * Structural deep clone.
 *
 * Only clones plain objects and arrays, other values are left alone.
 *
 * @param {(object|array)} obj
 * @return {(object|array)}
 */
export function structuralDeepClone(obj) {
  return structuralCloneImpl(obj, true, []);
}

/**
 * Structural shallow clone.
 *
 * Only clones plain objects and arrays, other values are left alone.
 *
 * @param {(object|array)} obj
 * @return {(object|array)}
 */
export function structuralShallowClone(obj) {
  return structuralCloneImpl(obj, false);
}

function structuralCloneImpl(obj, deep, stack) {
  if (stack && stack.includes(obj)) {
    throw new TypeError('Cannot clone circular structure.');
  }

  let result = obj;
  const isArr = Array.isArray(obj),
    isObj = isPlainObject(obj);

  if (isArr || isObj) {
    if (isArr) {
      result = [];
    } else {
      result = {};
    }

    for (let key in obj) {
      if (hasOwnProperty(obj, key)) {
        let substack;
        if (deep) {
          substack = stack.slice();
          substack.push(result);
        }
        result[key] = deep
          ? structuralCloneImpl(obj[key], deep, substack)
          : obj[key];
      }
    }
  }

  return result;
}

/**
 * Get the value of path at object.
 *
 * @param {object} object
 * @param {array} path Path of property to get, e.g. ['a', 2, 'q']
 * @returns {*} Value at path.
 */
export function get(object, path) {
  if (typeof path === 'string') {
    return object[path];
  }

  let i = 0;
  while (object != null && i < path.length) {
    object = object[path[i++]];
  }
  return object;
}

export const isIndex = val =>
  typeof val == 'number' || /^(?:0|[1-9]\d*)$/.test(val);

const setKey = (target, key, value) => {
  target[key] = value;
};

/**
 * Set the value at path of object.
 *
 * Arrays are created for missing index properties and objects are created for other missing properties.
 *
 * @param {object} object Object to modify.
 * @param {array} path Path of property to set, e.g. ['a', 2, 'q']
 * @param {*} value
 * @returns {object}
 */
export function set(object, path, value) {
  return baseSet(object, path, value, setKey);
}

/**
 * Base implementation for set and vue_util.set.
 *
 * @param {object} object Object to modify.
 * @param {array} path Path of property to set, e.g. ['a', 2, 'q']
 * @param {*} value
 * @param {*} setKey
 * @returns {object}
 */
export function baseSet(object, path, value, setKey) {
  if (typeof path === 'string') {
    setKey(object, path, value);
    return;
  }

  const lastIndex = path.length - 1;
  let index = 0;

  while (object != null && index <= lastIndex) {
    const key = path[index];
    if (index == lastIndex) {
      setKey(object, key, value);
    } else {
      if (object[key] == null) {
        // no object at key, create one
        setKey(object, key, isIndex(path[index + 1]) ? [] : {});
      }
    }
    object = object[key];
    index++;
  }
  return object;
}
