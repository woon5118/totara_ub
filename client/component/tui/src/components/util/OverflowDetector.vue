<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Simon Chester <simon.chester@totaralearning.com>
  @module tui
-->

<script>
import { throttle } from 'tui/util';
import { getOffsetRect } from 'tui/dom/position';
import ResizeObserver from 'tui/polyfills/ResizeObserver';
import { isRtl } from 'tui/i18n';

const THROTTLE_UPDATE = 150;

export default {
  data() {
    return {
      measuring: false,
      measured: false,
      overflowing: false,
      visible: Infinity,
    };
  },

  created() {
    // for detecting infinite loops
    this.iters = 0;
    this.failed = false;
  },

  mounted() {
    this.resizeObserver = new ResizeObserver(
      throttle(this.$_update, THROTTLE_UPDATE)
    );
    this.resizeObserver.observe(this.$el);
    this.$_update();
  },

  destroyed() {
    this.resizeObserver.disconnect();
  },

  updated() {
    this.$_update();
  },

  methods: {
    $_update() {
      if (!this.$el || this.failed) {
        return;
      }

      // infinite loop failsafe
      if (this.iters == 0) {
        setTimeout(() => (this.iters = 0), 10);
      }
      this.iters++;
      if (this.iters > 100) {
        console.error(
          '[OverflowDetector] Too many updates. Possible infinite loop, disabling.'
        );
        this.failed = true;
        this.measuring = true;
        return;
      }

      if (this.measuring) {
        // stage 2: measure
        this.$_measure();
        this.measuring = false;
        // mark as measured so we don't get in to an infinite loop with updated()
        this.measured = true;
      } else if (this.measured) {
        // stage 3: unset measured
        this.measured = false;
      } else {
        // stage 1: mark as measuring so we get all items shown
        this.measuring = true;
      }
    },

    /**
     * Check which child elements are overflowing and emit event if that has changed.
     */
    $_measure() {
      const el = this.$el;
      const pos = getOffsetRect(el);

      let overflowing = false;
      let visible = el.children.length;

      // figure out how many children are visible
      for (let i = 0; i < el.children.length; i++) {
        const child = el.children[i];
        const childPos = getOffsetRect(child);
        const overflowRTL = isRtl() ? childPos.left < pos.left - 1 : false;
        const overflowLTR = !isRtl() ? childPos.right > pos.right + 1 : false;
        if (
          overflowLTR ||
          overflowRTL ||
          childPos.bottom - pos.top - 1 > pos.height
        ) {
          // we hit our first non-visible child
          // i starts at 0, so it is equal to the number of visible children
          visible = i;
          overflowing = true;
          break;
        }
      }

      // emit event if state has changed
      if (this.overflowing != overflowing || this.visible != visible) {
        this.overflowing = overflowing;
        this.visible = visible;
        this.$emit('change', { overflowing, visible });
      }
    },
  },

  render() {
    return this.$scopedSlots.default({
      measuring: this.measuring,
    });
  },
};
</script>
