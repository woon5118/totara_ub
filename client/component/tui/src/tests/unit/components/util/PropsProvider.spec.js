/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import { mount } from '@vue/test-utils';
import { plainWrapperArray } from '../../util';
import PropsProvider from 'tui/components/util/PropsProvider';

const SecondLevel = {
  props: ['j'],
  render(h) {
    return h('span', {}, [`j: ${this.j}`]);
  },
};

const TestComp = {
  props: ['i', 'j'],
  render(h) {
    return h('p', {}, [`${this.i}, ${this.j}`, h(SecondLevel)]);
  },
};

const PropsProviderWrap = {
  functional: true,
  render(h, { props, children }) {
    return h('div', [h(PropsProvider, { props }, children)]);
  },
};

describe('presentation/util/PropsProvider.vue', () => {
  it('passes props and events to direct children', () => {
    const handler = jest.fn();
    const wrapper = mount(PropsProviderWrap, {
      context: {
        children: [
          h => h(TestComp, { props: { i: 1 } }),
          'hi',
          h => h(TestComp, { props: { i: 2 } }),
          h => h(TestComp, { props: { i: 3, j: 3 } }),
        ],
      },
      propsData: {
        provide({ props }) {
          return {
            props: { j: (props.i || 0) + 1000 },
            listeners: { click: handler },
          };
        },
      },
    });

    expect(
      plainWrapperArray(wrapper.findAll(TestComp)).map(w => w.props())
    ).toMatchObject([
      { i: 1, j: 1001 },
      { i: 2, j: 1002 },
      { i: 3, j: 3 },
    ]);
    expect(
      plainWrapperArray(wrapper.findAll(SecondLevel)).map(w => w.props())
    ).toMatchObject([{ j: undefined }, { j: undefined }, { j: undefined }]);

    expect(handler).toHaveBeenCalledTimes(0);
    wrapper.find(TestComp).vm.$emit('click', 'foo');
    expect(handler).toHaveBeenCalledWith('foo');

    expect(wrapper.element).toMatchSnapshot();
  });

  it('merges multiple event handlers', () => {
    const handlerInline = jest.fn();
    const handlerProvided = jest.fn();
    const wrapper = mount(PropsProviderWrap, {
      context: {
        children: [
          h => h(TestComp, { props: { i: 1 }, on: { click: handlerInline } }),
          h => h(TestComp, { props: { i: 2 }, on: { click: handlerInline } }),
          h => h(TestComp, { props: { i: 3 } }),
        ],
      },
      propsData: {
        provide({ props }) {
          return {
            props: { j: (props.i || 0) + 1000 },
            listeners: props.i != 2 ? { click: handlerProvided } : {},
          };
        },
      },
    });

    const comps = wrapper.findAll(TestComp);

    comps.at(0).vm.$emit('click', 'first');
    expect(handlerInline).toHaveBeenCalledWith('first');
    expect(handlerProvided).toHaveBeenCalledWith('first');

    comps.at(1).vm.$emit('click', 'second');
    expect(handlerInline).toHaveBeenCalledWith('second');
    expect(handlerProvided).not.toHaveBeenCalledWith('second');

    comps.at(2).vm.$emit('click', 'third');
    expect(handlerInline).not.toHaveBeenCalledWith('third');
    expect(handlerProvided).toHaveBeenCalledWith('third');
  });
});
