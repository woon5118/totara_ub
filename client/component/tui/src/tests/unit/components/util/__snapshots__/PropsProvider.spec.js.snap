// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presentation/util/PropsProvider.vue passes props and events to direct children 1`] = `
<div>
  <p>
    1, 1001
    <span>
      j: undefined
    </span>
  </p>
  hi
  <p>
    2, 1002
    <span>
      j: undefined
    </span>
  </p>
  <p>
    3, 3
    <span>
      j: undefined
    </span>
  </p>
</div>
`;
